### ISSUE TITLE
Clipboard doesn't sync in Wayland session

### ISSUE THREAD

**Issue by iota-dev** [andyholmes/gnome-shell-extension-gsconnect#20]
My Gnome shell version is 3.26, almost all features are working except clipboard sync.

**Comment by andyholmes** [comment]
Hmm, it works for me in both directions. Could you enabled "Debug Mode" in the About Page of preferences and tell me if there is any useful output related to the Clipboard plugin?

**Comment by iota-dev** [comment]
JS LOG: Daemon: Updating device
JS LOG: Daemon: Updating device
JS LOG: Authenticating 'Sony'
JS LOG: Connected to 'Sony'
JS LOG: error sending packet: Gio.IOErrorEnum: Error sending data: Broken pipe
JS LOG: error sending packet: Gio.TlsError: Error writing data to TLS socket: The specified session has been invalidated for some reason.
JS LOG: error sending packet: Gio.TlsError: Error writing data to TLS socket: The specified session has been invalidated for some reason.
JS LOG: Daemon: Updating device
JS LOG: Authenticating 'Sony'
JS LOG: error closing data output: Gio.IOErrorEnum: Error sending data: Broken pipe
JS LOG: [gsconnect@andyholmes.github.io]: LanChannel.receive(479cf10e4ea77d91)
JS LOG: [gsconnect@andyholmes.github.io]: Device received: {"id":1513537897644,"type":"kdeconnect.clipboard","body":{"content":"21294000711133066503"}}
JS LOG: [gsconnect@andyholmes.github.io]: Received from 'Sony'
JS LOG: [gsconnect@andyholmes.github.io]: Device.fromPacket(479cf10e4ea77d91)
JS LOG: [gsconnect@andyholmes.github.io]: Clipboard: handlePacket()
JS LOG: Error connecting: Gio.IOErrorEnum: Could not connect to 192.168.1.4: Connection timed out
JS LOG: [gsconnect@andyholmes.github.io]: LanChannel.receive(479cf10e4ea77d91)
JS LOG: [gsconnect@andyholmes.github.io]: Device received: {"id":1513537854008,"type":"kdeconnect.battery","body":{"currentCharge":48,"isCharging":false,"thresholdEvent":0}}
JS LOG: [gsconnect@andyholmes.github.io]: Received from 'Sony'
JS LOG: [gsconnect@andyholmes.github.io]: Device.fromPacket(479cf10e4ea77d91)
JS LOG: [gsconnect@andyholmes.github.io]: Battery: handlePacket()
JS LOG: [gsconnect@andyholmes.github.io]: Battery: receive()
JS LOG: [gsconnect@andyholmes.github.io]: LanChannel.receive(479cf10e4ea77d91)
JS LOG: [gsconnect@andyholmes.github.io]: Device received: {"id":1513538020604,"type":"kdeconnect.clipboard","body":{"content":"test clipboard"}}
JS LOG: [gsconnect@andyholmes.github.io]: Received from 'Sony'
JS LOG: [gsconnect@andyholmes.github.io]: Device.fromPacket(479cf10e4ea77d91)
JS LOG: [gsconnect@andyholmes.github.io]: Clipboard: handlePacket()

**Comment by iota-dev** [comment]
I'm using released zip package.

**Comment by andyholmes** [comment]
> JS LOG: [gsconnect@andyholmes.github.io]: Device received: {"id":1513538020604,"type":"kdeconnect.clipboard","body":{"content":"test clipboard"}}
> JS LOG: [gsconnect@andyholmes.github.io]: Received from 'Sony'

It looks like it's working, is the data not in your clipboard? Are you using any other clipboard managers that might be interfering?

**Comment by iota-dev** [comment]
Yes. I'm using "clipboard-indicator@tudmotu.com" gnome-shell extension.

**Comment by andyholmes** [comment]
Okay, there does seem to be a problem running along side that extension. Something weird happens and GSConnect receives the clipboard content, but sends it right back instead of adding it to the clipboard.

I'll see if I can find a workaround and get back to you.

**Comment by iota-dev** [comment]
Thank you.
In kdeconnect-indicator clipboard work fine alongside clipboard-indicator extension.

**Comment by andyholmes** [comment]
That's good to know, I'll have a look at how kdeconnect does it.

**Comment by andyholmes** [comment]
After some more investigating, the plugin still seems to be working for me even with clipboard-indicator running. I can even copy on my tablet, tablet forwards to laptop, laptop forwards to phone.

**Comment by iota-dev** [comment]
But not working for me. although after deactivate clipboard-indicator.

**Comment by andyholmes** [comment]
You can try this Zip, it tries not to send duplicate content to devices. But I not sure what the problem is for sure, since I can reproduce it.

[gsconnect@andyholmes.github.io.zip](https://github.com/andyholmes/gnome-shell-extension-gsconnect/files/1566267/gsconnect.andyholmes.github.io.zip)

**Comment by andyholmes** [comment]
I still haven't been able to reproduce this and no other users have reported or voted for it, so I'm going to close it until more information is available or other users come forward.

**Comment by lnicola** [comment]
Is clipboard sync supposed to work on Wayland? I have the same issue.

**Comment by andyholmes** [comment]
There are some differences between Wayland and Xorg regarding the clipboard, they're summed up pretty well [here](https://wiki.gnome.org/Initiatives/Wayland/PrimarySelection). I don't use Wayland much myself (eg. I try it once in awhile to see how it's progressing) so I don't know what the state of the clipboard situation is currently, or whether certain distros like Ubuntu make any attempts to smooth these problems by default.

Could you tell me if the clipboard problem is present for both PRIMARY (eg, mouse middle-click) and CLIPBOARD (eg. <kbd>Ctrl</kbd> + <kbd>C</kbd>) situations?

**Comment by andyholmes** [comment]
Okay, I've started up Wayland and this is indeed broken in that session-type. I believe I've found a workaround, by way of GPaste, which forces use of X11 for the Gdk backend. If you could give this Zip a good run through, I'd appreciate it, since I'm unsure if this will break anything else:

[gsconnect@andyholmes.github.io.zip](https://github.com/andyholmes/gnome-shell-extension-gsconnect/files/1822303/gsconnect.andyholmes.github.io.zip)

**Comment by lnicola** [comment]
Thanks, the new version seems to be working in both directions.

I had to pair again with my phone, but that's probably a different issue (its IP might have changed? though there was no reason to).

About Wayland, I've been using it for quite some time on Arch and it's pretty usable. Some things are slower (no HW acceleration for video decoding, [sometimes sluggish](https://bugzilla.gnome.org/show_bug.cgi?id=745032)), the compositor can't be restarted and bad extensions  and GJS issues bring down the whole session (although that didn't happen to me in a while). Other than that it's fine :smile:.

**Comment by andyholmes** [comment]
> Thanks, the new version seems to be working in both directions.

Excellent, I had to do some digging on this. Apparently in wayland there is no such thing as desktop-wide clipboard access (for securirty), unless someone writes a compositor extension for it. Sadly, that means this might disappear at some point, but we'll see.

> I had to pair again with my phone, but that's probably a different issue (its IP might have changed? though there was no reason to).

Probably you falling victim to #26 which was actually a lot worse of a bug than I first thought, since apparently KDE Connect doesn't handle missing deviceId's, *at all* (I have to remember to file an upstream bug about that, since it could be a security concern). It basically tricked KDE Connect clients into thinking they were all the same. The fix for that happens to be in the build I posted here as well.

> About Wayland, I've been using it for quite some time...

You might be happy to hear I "overheard" this in #gnome-shell on gimpNet today:

> feaneron (IRC)
> HA!
> i managed to fix the super annoying Wayland stuttering on dual-GPU laptops!!!!

https://gitlab.gnome.org/GNOME/mutter/merge_requests/50
